File: src/main/scala/info/crlog/higgs/protocol/boson/BosonMessage.scala
package info.crlog.higgs.protocol.boson

import info.crlog.higgs.protocol.Message
import reflect.BeanProperty
import info.crlog.higgs.util.StringUtil

/**
 * @author Courtney Robinson <courtney@crlog.info> @ 31/01/12
 */

class BosonMessage extends Message {

  object FLAGS {
    /**
     * Full message with no more content
     */
    val NO_MORE_CONTENT: Byte = 0x0
    /**
     * Multi-part message which is delivered as a single message on the receiving end
     */
    val MULTI_PART_MESSAGE: Byte = 0x1
    /**
     * Split Multi-part message which delivers each part of a multi part message separately on the receiving end.
     */
    val SPLIT_MULTI_PART_MESSAGE: Byte = 0x2
    /**
     * only valid when used with split multi part message
     * 's' split multi part messages are no longer delivered at the end once all parts of the
     * message is received, instead each part of the message is delivered, without buffering
     */
    val NO_MESSAGE_BUFFER: Byte = 0x3

  }

  def this(message: String) = {
    this()
    contents = message.getBytes
  }

  /**
   * Create a message supplying its topic
   * @param t the topic of this message
   * @param c the contents of the message
   */
  def this(t: String, c: String) = {
    this()
    topic = t
    contents = c.getBytes
  }

  def this(bytes: Array[Byte]) = {
    this()
    contents = bytes
  }

  /**
   * Create a message supplying its flag,topic and message content
   * @param f   the flag for this message, must be one of the constants in BosonMessage.FLAGS.*
   * @param topic  the topic of the message, a topic's length is limited to 32,767 bytes, about 30KB i.e. the MAX value of a short
   * @param content the contents of the message
   */
  def this(f: Byte, topic: Array[Byte], content: Array[Byte]) = {
    this()
    flag = f
    this.topic = StringUtil.getString(topic)
    contents = content
  }

  /**
   * Simple,naive constructor which simply calls toString on the provided object
   * i.e. your object must override toString and return the string form you wish to be sent
   */
  def this(obj: AnyRef) = {
    this()
    if (obj.isInstanceOf[Array[Byte]]) {
      contents = obj.asInstanceOf[Array[Byte]]
    } else {
      contents = obj.toString.getBytes
    }
  }

  @BeanProperty
  var flag: Byte = FLAGS.NO_MORE_CONTENT
}

File: src/main/java/info/crlog/higgs/protocol/boson/BosonMessage.java
package info.crlog.higgs.protocol.boson;

import info.crlog.higgs.messaging.ReUsableMessage;
import info.crlog.higgs.util.StringUtil;
import io.netty.buffer.ChannelBuffer;
import io.netty.buffer.ChannelBuffers;

/**
 *
 * @author Courtney Robinson <courtney@crlog.info>
 */
public class BosonMessage implements ReUsableMessage {

    protected StringUtil util = new StringUtil();
    protected String topic = "";
    protected String content = "";
    private short version = 2;

    public BosonMessage() {
    }

    public BosonMessage(String topic, String content) {
        this.topic = topic;
        this.content = content;
    }

    public BosonMessage(String content) {
        this.content = content;
    }

    public void clean() {
        topic = "";
        content = "";
    }

    public int getMessageSize(ChannelBuffer buffer) {
        // Wait until the protocol version and message size is available
        if (buffer.readableBytes() < 6) {
            return -1;
        }
        version = buffer.readShort(); //protocol version, first 2 bytes
        return buffer.readInt(); //message size, 3rd to 6th bytes, i.e. 32 bits
    }

    public void deserialize(ChannelBuffer buf) {
        //put the protocol version
        version = buf.readShort();
        //get the message length
        buf.readInt();
        //de-serialize the topic
        //get the property key's size
        short topicKeyLength = buf.readShort();
        if (topicKeyLength > 0) //get the property key
        {
            buf.readBytes(topicKeyLength);
        }
        //get the property value's size
        int topicValueLength = buf.readInt();
        if (topicValueLength > 0) //get the property value
        {
            topic = util.getString(buf.readBytes(topicValueLength).array());
        }
        //now de-serialize the message contents
        //get the property key's size
        short valueKeyLength = buf.readShort();
        if (valueKeyLength > 0) //add the property key
        {
            buf.readBytes(valueKeyLength);
        }
        //get the property value's size
        int valueLength = buf.readInt();
        if (valueLength > 0) //get the property value
        {
            content = util.getString(buf.readBytes(valueLength).array());
        }
    }

    public ChannelBuffer serialize() {
        ChannelBuffer buf = ChannelBuffers.dynamicBuffer();
        //put the protocol version
        buf.writeShort(version);
        //serialize topic and conent...
        String topicStr = "topic";
        String contentStr = "content";
        byte[] topicKey = util.getBytes(topicStr);
        byte[] topicContent = util.getBytes(topic);
        byte[] valueKey = util.getBytes(contentStr);
        byte[] valueContent = util.getBytes(content);
        //write the message length
        int messageSize = topicKey.length + topicContent.length
                + valueKey.length + valueContent.length;
        buf.writeInt(messageSize);
        //serialize the topic
        //add the property key's size
        buf.writeShort(topicKey.length);
        //add the property key
        buf.writeBytes(topicKey);
        //add the property value's size
        buf.writeInt(topicContent.length);
        //add the property value
        buf.writeBytes(topicContent);

        //now serialize the message contents
        //add the property key's size
        buf.writeShort(valueKey.length);
        //add the property key
        buf.writeBytes(valueKey);
        //add the property value's size
        buf.writeInt(valueContent.length);
        //add the property value
        buf.writeBytes(valueContent);

        return buf;
    }

    public BosonMessage setContent(String msg) {
        content = msg;
        return this;
    }

    public BosonMessage setTopic(String topic) {
        this.topic = topic;
        return this;
    }

    @Override
    public String toString() {
        return "t:" + topic + "<->c:" + content;
    }

    public String getTopic() {
        return topic;
    }

    public String getContent() {
        return content;
    }
}
